pymeasure/adapters/adapter.py

```diff
@@ -41,37 +43,75 @@ class Adapter:
     def __init__(self, preprocess_reply=None, **kwargs):
         self.preprocess_reply = preprocess_reply
         self.connection = None
+        self.log = logging.Logger("Adapter")
+        self.log.addHandler(logging.NullHandler())
 
     def __del__(self):
         """Close connection upon garbage collection of the device"""
         if self.connection is not None:
             self.connection.close()
 
+    # Directly called methods.
     def write(self, command):
-        """ Writes a command to the instrument
+        """
+        Write a command to the instrument.
 
         :param command: SCPI command string to be sent to the instrument
         """
-        raise NameError("Adapter (sub)class has not implemented writing")
+        self.log.debug(f"WRITE:{command.encode()}")
+        self._write(command)
+
+    def write_bytes(self, content):
+        """Write the bytes `content` to the instrument."""
+        self.log.debug(f"WRITE:{content}")
+        self._write_bytes(content)
+
+    def read(self):
+        """
+        Read until the buffer is empty and return the resulting string.
+
+        :returns: String ASCII response of the instrument.
+        """
+        read = self._read()
+        self.log.debug(f"READ:{read.encode()}")
+        return read
+
+    def read_bytes(self, count):
+        """Read `count` number of bytes from the instrument."""
+        read = self._read_bytes(count)
+        self.log.debug(f"READ:{read}")
+        return read
+
+    # Methods to implement in the subclass.
 
+    def _write(self, command):
+        """Write to the instrument. Implement in subclass."""
+        raise NotImplementedError("Adapter class has not implemented writing")
+
+    def _write_bytes(self, content):
+        """Write bytes to the instrument. Implement in subclass."""
+        raise NotImplementedError("Adapter class has not implemented writing")
+
+    def _read(self):
+        """Read from the instrument. Implement in subclass."""
+        raise NotImplementedError("Adapter class has not implemented reading")
+
+    def _read_bytes(self, count):
+        """Read from the instrument. Implement in subclass."""
+        raise NotImplementedError("Adapter class has not implemented reading")
+
+    # Deprecated methods.
     def ask(self, command):
         """ Writes the command to the instrument and returns the resulting
         ASCII response
 
         :param command: SCPI command string to be sent to the instrument
         :returns: String ASCII response of the instrument
         """
+        raise DeprecationWarning("Ask is now in the instrument")
```

Comment:
Two points:
* when encounering deprecation warnings, I always appreciate it when the error message clearly spells out what I need to do to fix the problem. Here, I'm not sure if it's 100% clear.
* For whom is this warning intended? I'm asking because (not sure if you're aware) DeprecationWarnings by default are filtered out by the interpreter, so you _only see them_ with some settings (e.g. invoking python with ` -X dev`). If everybody should see this if it gets hit (and hopefully report it), use `FutureWarning` instead.